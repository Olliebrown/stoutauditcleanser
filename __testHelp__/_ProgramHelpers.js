import fs from 'fs'
import path from 'path'

// Project specific imports
import { QUERIES } from '../src/domTraversal/queriesAndRegex.js'
import Program from '../src/Objects/Program.js'

export function InitProgramTest (htmlFileName) {
  // Read in the HTML page
  const baseHTML = fs.readFileSync(
    path.join('__tests__', 'data', 'ProgramTables', htmlFileName),
    'utf8'
  )

  // Load page from imported HTML
  document.documentElement.innerHTML = baseHTML.toString()

  // Extract program headers and make sure there's at least one
  const headerNodes = document.querySelectorAll(QUERIES.programHeader)
  if (headerNodes?.length < 1) {
    throw new Error('Unable to find at least one program')
  }

  // Create program object
  return new Program(headerNodes[0])
}
