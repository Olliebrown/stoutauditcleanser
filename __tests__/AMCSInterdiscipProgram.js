// Necessary when using ECMA modules
import { jest } from '@jest/globals'

// Help load the program HTML page
import { InitProgramTest } from '../__testHelp__/_ProgramHelpers.js'

// Disable jest mocking for fs module
jest.dontMock('fs')

describe('BS in AMCS: Interdisciplinary Program Node', () => {
  // Inject clean version of HTML page and rebuild program before tests
  let program = null
  beforeAll(() => {
    // Load page from imported HTML
    program = InitProgramTest('AMCSInterdiscipProgram.html')
  })

  // Clear out the HTML and program
  afterAll(() => {
    jest.resetModules()
    program = null
  })

  // Basic page structure tests
  test('should have a specific page title and specific program prefix', () => {
    // Check title & program name
    const titleElem = document.querySelector('title')
    expect(titleElem?.textContent).toBe('BS Applied Math Computer Science Interdisciplinary')
    expect(program?.getName()).toMatch(/^BS In AMCS: interdisciplinary CONC\s+\d\d\d\d$/)
  })

  test('should have expected snapshot', () => {
    expect(program?.toJSON()).toMatchSnapshot()
  })
})
