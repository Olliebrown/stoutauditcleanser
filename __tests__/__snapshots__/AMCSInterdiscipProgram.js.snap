// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BS in AMCS: Interdisciplinary Program Node should have expected snapshot 1`] = `
{
  "internalId": "RG-10776",
  "isGenEd": false,
  "key": "bs_in_amcs__interdisciplinary_conc_2020",
  "name": "BS In AMCS: interdisciplinary CONC 2020",
  "satisfied": "IN_PROGRESS",
  "subNodes": [
    {
      "description": "Racial and Ethnic StudiesStudents must take 6 credits, a minimum of 3 credits must be from RES-A",
      "index": 1,
      "internalId": "RQ-10360",
      "isGenEd": true,
      "key": "res_requirement",
      "name": "RES Requirement",
      "nextIndex": 5,
      "satisfied": "COMPLETE",
      "subNodes": [
        {
          "description": "Option 1: Take two RES-A Course",
          "internalId": "LN-10",
          "isGenEd": true,
          "isValid": true,
          "key": "option_1_take_two_res_a_course",
          "name": "Option 1:take Two RES-a Course",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
      ],
      "units": {
        "credits": {
          "current": 6,
          "required": 6,
        },
      },
    },
    {
      "description": "Global Perspectives",
      "index": 5,
      "internalId": "RQ-025",
      "isGenEd": true,
      "key": "glp_requirement",
      "name": "GLP Requirement",
      "nextIndex": 9,
      "satisfied": "COMPLETE",
      "subNodes": [
        {
          "description": "Global Perspective Courses  - 6 Credits",
          "internalId": "LN-010",
          "isGenEd": false,
          "isValid": true,
          "key": "global_perspective_courses",
          "name": "Global Perspective Courses",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
      ],
      "units": {
        "credits": {
          "current": 9,
          "required": 6,
        },
      },
    },
    {
      "description": "General Education: Communication Skills",
      "index": 9,
      "internalId": "RQ-020",
      "isGenEd": true,
      "key": "gened__communication_skills",
      "name": "GenEd: Communication Skills",
      "nextIndex": 17,
      "satisfied": "COMPLETE",
      "subNodes": [
        {
          "description": "Take ENGL-101 or ENGL-111HON",
          "internalId": "LN-010",
          "isGenEd": false,
          "isValid": true,
          "key": "freshman_composition_1___engl_101_or_engl_111hon",
          "name": "Freshman Composition 1 - ENGL-101 Or ENGL-111hon",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "Take ENGL-102 or ENGL-113HON",
          "internalId": "LN-020",
          "isGenEd": false,
          "isValid": true,
          "key": "freshman_composition_2___engl_102_or_engl_113hon",
          "name": "Freshman Composition 2 - ENGL-102 Or ENGL-113hon",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "Take COMST 100",
          "internalId": "LN-030",
          "isGenEd": false,
          "isValid": true,
          "key": "fundamentals_of_speech___comst_100",
          "name": "Fundamentals Of Speech - COMST 100",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
      ],
      "units": {},
    },
    {
      "description": "General Education: Analytic Reasoning & Natural Science.1Take MATH-156 or MATH-153*MATH 156 and MATH 153 require a grade of C or better",
      "index": 17,
      "internalId": "RQ-10415",
      "isGenEd": true,
      "key": "gened__arns",
      "name": "GenEd: ARNS",
      "nextIndex": 17,
      "satisfied": "COMPLETE",
      "subNodes": [],
      "units": {},
    },
    {
      "description": "General Education: Analytic Reasoning & Natural Science.2",
      "index": 23,
      "internalId": "RQ-10416",
      "isGenEd": true,
      "key": "gened__arns",
      "name": "GenEd: ARNS",
      "nextIndex": 33,
      "satisfied": "COMPLETE",
      "subNodes": [
        {
          "description": "Take MATH-157",
          "internalId": "LN-10",
          "isGenEd": false,
          "isValid": true,
          "key": "math_157",
          "name": "MATH-157",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "LN- 20: Take a Natural Sciences course from the approved Gen Ed/Stout Core-ARNS Natural Science course listing which can be found online",
          "internalId": "UNKNOWN_ID",
          "isGenEd": false,
          "isValid": true,
          "key": "natural_sciences",
          "name": "Natural Sciences",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "Laboratory Course Required",
          "internalId": "LN-30",
          "isGenEd": false,
          "isValid": true,
          "key": "laboratory_course_required",
          "name": "Laboratory Course Required",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
      ],
      "units": {
        "courses": {
          "current": 1,
          "required": 1,
        },
        "credits": {
          "current": 9,
          "required": 8,
        },
      },
    },
    {
      "description": "General Education: Humanities and the Arts- Select from two or more areas Including: Art History/Music Appreciation, Creative/Performing Arts, History, Literature, Philosophy, & Foreign Language & Culture  *Effective 01/01/2021, PHOTO 130 NO LONGER classified as GE HUM/ART & does not satisfy this area.   **Effective 08/15/2021, ART 100 & ART 101 NO LONGER classified as GE HUM/ART & do not satisfy this area.",
      "index": 33,
      "internalId": "RQ-019",
      "isGenEd": true,
      "key": "gened__humart",
      "name": "GenEd: HumArt",
      "nextIndex": 41,
      "satisfied": "COMPLETE",
      "subNodes": [
        {
          "description": "Creative/Performing Arts Option ",
          "internalId": "LN-030",
          "isGenEd": false,
          "isValid": true,
          "key": "creative_performing_arts_option",
          "name": "Creative/performing Arts Option",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "History Option ",
          "internalId": "LN-070",
          "isGenEd": false,
          "isValid": true,
          "key": "history_option",
          "name": "History Option",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
      ],
      "units": {
        "credits": {
          "current": 6,
          "required": 6,
        },
      },
    },
    {
      "description": "General Education: Social and Behavioral Sciences - Courses MUST be from two or more Areas Including: Anthropology, Economics, Geography, Political Science, Psychology and Sociology",
      "index": 41,
      "internalId": "RQ-0138",
      "isGenEd": true,
      "key": "gened__sbsci",
      "name": "GenEd: SBSci",
      "nextIndex": 47,
      "satisfied": "COMPLETE",
      "subNodes": [
        {
          "description": "Anthropology Option ",
          "internalId": "LN-010",
          "isGenEd": false,
          "isValid": true,
          "key": "anthropology",
          "name": "Anthropology",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
      ],
      "units": {
        "credits": {
          "current": 3,
          "required": 6,
        },
      },
    },
    {
      "description": "General Education: Cross-Disciplinary Issues",
      "index": 47,
      "internalId": "RQ-10361",
      "isGenEd": true,
      "key": "gened__ci",
      "name": "GenEd: CI",
      "nextIndex": 51,
      "satisfied": "COMPLETE",
      "subNodes": [
        {
          "description": "Complete 3 credits from the approved CISS course list",
          "internalId": "LN-010",
          "isGenEd": false,
          "isValid": true,
          "key": "select_3_credits",
          "name": "Select 3 Credits",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
      ],
      "units": {
        "credits": {
          "current": 3,
          "required": 3,
        },
      },
    },
    {
      "description": "General Education: Social Responsibility and Ethical Reasoning",
      "index": 51,
      "internalId": "RQ-10362",
      "isGenEd": true,
      "key": "gened__srer",
      "name": "GenEd: SRER",
      "nextIndex": 55,
      "satisfied": "COMPLETE",
      "subNodes": [
        {
          "description": "Complete 3 credits from the approved Gen Ed/Stout Core-SRER course listing which can be found online.",
          "internalId": "LN-010",
          "isGenEd": false,
          "isValid": true,
          "key": "take_3_credits",
          "name": "Take 3 Credits",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
      ],
      "units": {
        "credits": {
          "current": 3,
          "required": 3,
        },
      },
    },
    {
      "description": "GENERAL EDUCATION: SELECTIVES",
      "index": 55,
      "internalId": "RQ-10414",
      "isGenEd": true,
      "key": "gened__selectives",
      "name": "GenEd: Selectives",
      "nextIndex": 61,
      "satisfied": "COMPLETE",
      "subNodes": [
        {
          "description": "Take CS-144",
          "internalId": "LN-010",
          "isGenEd": false,
          "isValid": true,
          "key": "take_cs_144",
          "name": "Take CS-144",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
      ],
      "units": {
        "credits": {
          "current": 4,
          "required": 4,
        },
        "gpa": {
          "current": 0,
          "required": 0,
        },
      },
    },
    {
      "description": "Applied Mathematics & Computer Science Major Requirements Interdisciplinary Concentration",
      "index": 61,
      "internalId": "RQ-0384",
      "isGenEd": false,
      "key": "amcs__interdisciplinary__requirements",
      "name": "AMCS: interdisciplinary  Requirements",
      "nextIndex": 81,
      "satisfied": "IN_PROGRESS",
      "subNodes": [
        {
          "description": "Mathematics Requirements",
          "internalId": "LN-010",
          "isGenEd": false,
          "isValid": true,
          "key": "mathematics_requirements___part_1",
          "name": "Mathematics Requirements - Part 1",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "Take MATH 180 or MATH 270MATH 270 is one credit less. If this course is taken, the additional credit must be made up in Selectives Area per blanket waiver approved D. Ding 11/18/2021",
          "internalId": "LN-015",
          "isGenEd": false,
          "isValid": true,
          "key": "math_requirements___part_2",
          "name": "MATH Requirements - Part 2",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "Take STAT 330 and choose 6 credits from the following courses:",
          "internalId": "LN-20",
          "isGenEd": false,
          "isValid": true,
          "key": "statistics_analysis___option_1",
          "name": "Statistics/analysis - Option 1",
          "satisfied": "INCOMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "Take STAT 331 AND 332 and choose 3 credits from the following courses:",
          "internalId": "LN-030",
          "isGenEd": false,
          "isValid": true,
          "key": "statistics_analysis___option_2",
          "name": "Statistics/analysis - Option 2",
          "satisfied": "INCOMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "Complete CS 145, CS 244, CS 324 and 6 credits of MSCS 200 or higher and/or CS 246 or higherCS 145 and CS 244 require a grade of C or higher",
          "internalId": "LN-040",
          "isGenEd": false,
          "isValid": true,
          "key": "cs___mscs_req",
          "name": "CS & MSCS Req",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "Cooperative Educational Experience",
          "internalId": "LN-045",
          "isGenEd": false,
          "isValid": true,
          "key": "cooperative_educational_experience",
          "name": "Cooperative Educational Experience",
          "satisfied": "INCOMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "Selective Courses",
          "internalId": "LN-60",
          "isGenEd": false,
          "isValid": true,
          "key": "selective_courses",
          "name": "Selective Courses",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "Required Credits in an Approved Related Field and ElectivesProposed program plan must be approved by Program Director at least 4 semesters prior to graduation. Plan must indicate career path and proposed course to support that career. Course plan may include a minor or a second major.",
          "internalId": "LN-070",
          "isGenEd": false,
          "isValid": true,
          "key": "required_credits_in_an_approved_related_field___electives",
          "name": "Required Credits In An Approved Related Field & Electives",
          "satisfied": "INCOMPLETE",
          "subNodes": [],
          "units": {},
        },
      ],
      "units": {
        "courses": {
          "current": 3,
          "required": 3,
        },
        "credits": {
          "current": 50,
          "required": 77,
        },
      },
    },
    {
      "description": "OTHER COURSES",
      "index": 81,
      "internalId": "RQ-0331",
      "isGenEd": false,
      "key": "other_courses",
      "name": "Other Courses",
      "nextIndex": 86,
      "satisfied": "COMPLETE",
      "subNodes": [
        {
          "description": "Courses Not Used Toward Degree",
          "internalId": "UNKNOWN_ID",
          "isGenEd": false,
          "isValid": true,
          "key": "courses_not_used_toward_degree",
          "name": "Courses Not Used Toward Degree",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "Transfer Courses Not Used Toward Degree",
          "internalId": "UNKNOWN_ID",
          "isGenEd": false,
          "isValid": true,
          "key": "transfer_courses_not_used_toward_degree",
          "name": "Transfer Courses Not Used Toward Degree",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
      ],
      "units": {},
    },
  ],
  "units": {
    "credits": {
      "current": 98,
      "required": 120,
    },
    "gpa": {
      "current": 2.912,
      "required": 2,
    },
  },
}
`;
